examples/rl/policy_gradient/rl_policy_gradient_solver.py

```diff
@@ -0,0 +1,146 @@
+        .forward(X)
+        .choose_action(p)
+        .loss(xs, ys, rs)
+        .discount_rewards(rs)
+        .preprocessor
+    """
+    def __init__(self, model, env, **kwargs):
+        self.model = model
+        self.env = env
+        self.update_every = kwargs.pop('update_every', 10)
+        self.num_episodes = kwargs.pop('num_episodes', 100000)
+        self.save_every = kwargs.pop('save_every', 10)
+        self.save_dir = kwargs.pop('save_dir', '')
+        self.resume_from = kwargs.pop('resume_from', None)
+        self.render = kwargs.pop('render', False)
+
+        self.running_reward = None
+        self.episode_reward = 0
+
+        super(RLPolicyGradientSolver, self).__init__(model, None, None, **kwargs)
+
+    def _reset_data_iterators(self):
+        # This `Solver` does not use data iterators.
+        pass
+
+    def init(self):
+        if not os.path.exists(self.save_dir):
+            os.makedirs(self.save_dir)
+
+        if self.resume_from is not None:
+            with open(self.resume_from) as f:
+                params = pickle.load(f)
+            self.model.params = {k: np.array(v.tolist()) for k, v in params.iteritems()}
+        else:
+            super(RLPolicyGradientSolver, self).init()
+
+    def run_episode(self):
+        """Run an episode (multiple games) using the current model to generate training data.
+        :return tuple (xs, ys, rs): The N x input_size observations, N x 1 action labels,
+                and N x 1 discounted rewards obtained from running the episode's N steps.
+        """
+        observation = self.env.reset()
+        self.model.preprocessor.reset()
+        self.episode_reward = 0
+
+        xs, ys, rs = [], [], []
+        done = False
+        game_number = 1
+        game_start = time.time()
+        while not done:
+            if self.render:
+                self.env.render()
+            x = self.model.preprocessor.preprocess(observation)
+            p = self.model.forward(x)
+            a, y = self.model.choose_action(p.asnumpy().ravel()[0])
+            observation, r, done, info = self.env.step(a)
+
+            xs.append(x.asnumpy().ravel())
+            ys.append(y)
+            rs.append(r)
+            self.episode_reward += r
+            if self._game_complete(r):
+                game_time = time.time() - game_start
+                if self.verbose:
+                    print('game %d complete (%.2fs), reward: %f' % (game_number, game_time, r))
+                game_number += 1
+                game_start = time.time()
+
+        # Episode finished.
+        self.running_reward = self.episode_reward if self.running_reward is None else (
+            0.99*self.running_reward + 0.01*self.episode_reward)
+        xs = np.vstack(xs)
+        ys = np.vstack(ys)
+        rs = np.expand_dims(self.model.discount_rewards(rs), axis=1)
+        return xs, ys, rs
+
+    def _game_complete(self, reward):
+        return reward != 0
+
+    def train(self):
```

Comment:
Could you add more comments for this function? Especially the idea of two forward passes (one in loss definition, one in episode). This is really some interesting points to walkaround some limitation of dump backward path. Thanks!